### PR TITLE
[BEAM-2039] Number programming guide chapters

### PR DESCRIPTION
**Pull request by melap**
Includes the following consistency/formatting changes:
- Number chapters of programming guide (Re: BEAM-2039). I couldn't figure out how to make the table of contents use 1. 1.1. 1.1.1 etc. for each page, so I have it pulling from the section titles but it unfortunately still has the bullets. any ideas on how to do this better are welcome.
- Make capitalization of section titles consistent
- Format text to be more consistent, reasonable line length (no more 500+ character lines)
- Use the consistent automatic anchor link naming based on section title instead of making our own for arbitrary sections. the auto-generated toc did not work with our hard coded anchors either.

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/beam_PreCommit_Website_Stage/701/

Jenkins built the site at commit id 9c2e36bb7848b703fa46165e37d0a482c517c0a3 with Jekyll and staged it [here](http://apache-beam-website-pull-requests.storage.googleapis.com/320/index.html). Happy reviewing.

Note that any previous site has been deleted. This staged site will be automatically deleted after its TTL expires. Push any commit to the pull request branch or re-trigger the build to get it staged again.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/beam_PreCommit_Website_Stage/702/

Jenkins built the site at commit id 7d681499a1c36441bbd870d4b0f49ad04748ea84 with Jekyll and staged it [here](http://apache-beam-website-pull-requests.storage.googleapis.com/320/index.html). Happy reviewing.

Note that any previous site has been deleted. This staged site will be automatically deleted after its TTL expires. Push any commit to the pull request branch or re-trigger the build to get it staged again.

**Comment by melap** [comment]
R: @kennknowles 
CC: @XuMingmin

File: src/documentation/programming-guide.md [hunk 1]
```diff
@@ -1710,21 +2501,27 @@ If our trigger is set to `.accumulatingFiredPanes`, the trigger emits the follow
 ```
 
 
-##### Discarding Mode
+##### 8.4.1.2. Discarding mode:
```
Comment (tgroh) [inline src/documentation/programming-guide.md]: aaaaaaaaaaaaaaaaaaaaaa

File: src/documentation/programming-guide.md [hunk 2]
```diff
@@ -641,62 +936,83 @@ public class AverageFn extends CombineFn<Integer, AverageFn.Accum, Double> {
...
-* Specify `.asSingletonView`, in which the output is immediately converted to a `PCollectionView`, which will provide a default value for each empty window when used as a side input. You'll generally only need to use this option if the result of your pipeline's `Combine` is to be used as a side input later in the pipeline.
+* Specify `.withoutDefaults`, where windows that are empty in the input
+  `PCollection` will likewise be empty in the output collection.
+* Specify `.asSingletonView`, in which the output is immediately converted to a
+  `PCollectionView`, which will provide a default value for each empty window
+  when used as a side input. You'll generally only need to use this option if
+  the result of your pipeline's `Combine` is to be used as a side input later in
+  the pipeline.
 
+##### 4.2.4.6. Combining values in a key-grouped collection
```
Comment (tgroh) [inline src/documentation/programming-guide.md]: 'in a keyed PCollection'?

File: src/documentation/programming-guide.md [hunk 3]
```diff
@@ -1010,26 +1421,41 @@ While `ParDo` always produces a main output `PCollection` (as the return value f
...
+transform can make your code more modular and easier to understand.
 
-The Beam SDK comes packed with many useful composite transforms. See the API reference pages for a list of transforms:
+The Beam SDK comes packed with many useful composite transforms. See the API
+reference pages for a list of transforms:
   * [Pre-written Beam transforms for Java]({{ site.baseurl }}/documentation/sdks/javadoc/{{ site.release_latest }}/index.html?org/apache/beam/sdk/transforms/package-summary.html)
   * [Pre-written Beam transforms for Python]({{ site.baseurl }}/documentation/sdks/pydoc/{{ site.release_latest }}/apache_beam.transforms.html)
 
-### An example of a composite transform
+#### 4.6.1. Composite transform example
```
Comment (tgroh) [inline src/documentation/programming-guide.md]: I think this is too robotic and the previous is too verbose; maybe

'An example [composite ]transform'

File: src/documentation/programming-guide.md [hunk 4]
```diff
@@ -1095,24 +1531,41 @@ The following code sample shows how to override `expand` for the `ComputeWordLen
...
+directly by the user of a transform. Instead, you should call the `apply` method
+on the `PCollection` itself, with the transform as an argument. This allows
+transforms to be nested within the structure of your pipeline.
 
-## <a name="io"></a>Pipeline I/O
+#### 4.6.3. PTransform Style Guide
 
-When you create a pipeline, you often need to read data from some external source, such as a file in external data sink or a database. Likewise, you may want your pipeline to output its result data to a similar external data sink. Beam provides read and write transforms for a [number of common data storage types]({{site.baseurl }}/documentation/io/built-in/). If you want your pipeline to read from or write to a data storage format that isn't supported by the built-in transforms, you can [implement your own read and write transforms]({{site.baseurl }}/documentation/io/io-toc/).
+When you create a new `PTransform`, be sure to read the [PTransform Style
+Guide]({{ site.baseurl }}/contribute/ptransform-style-guide/). The guide
```
Comment (tgroh) [inline src/documentation/programming-guide.md]: So demanding!

Maybe "The PTransform style guide contains additional information not included here such as...", "It is a useful starting point when writing new composite PTransforms"?

File: src/documentation/programming-guide.md [hunk 5]
```diff
@@ -1372,119 +1923,228 @@ public class MyCustomDataType {
...
-As before, the pipeline creates a bounded `PCollection` of key/value pairs. We then set a [windowing function](#setwindowingfunction) for that `PCollection`. The `GroupByKey` transform groups the elements of the `PCollection` by both key and window, based on the windowing function. The subsequent `ParDo` transform gets applied multiple times per key, once for each window.
+As before, the pipeline creates a bounded `PCollection` of key/value pairs. We
+then set a [windowing function](#setting-your-pcollections-windowing-function)
+for that `PCollection`.  The `GroupByKey` transform groups the elements of the
+`PCollection` by both key and window, based on the windowing function. The
+subsequent `ParDo` transform gets applied multiple times per key, once for each
+window.
 
-### Windowing functions
+### 7.2. Beam windowing functions
```
Comment (tgroh) [inline src/documentation/programming-guide.md]: "Provided/Built-in window functions"?

File: src/documentation/programming-guide.md [hunk 6]
```diff
@@ -1372,119 +1923,228 @@ public class MyCustomDataType {
...
+the start of a new window.
 
 ![Diagram of session windows with a minimum gap duration]({{ "/images/session-windows.png" | prepend: site.baseurl }} "Session windows, with a minimum gap duration")
 
-**Figure:** Session windows, with a minimum gap duration. Note how each data key has different windows, according to its data distribution.
+**Figure:** Session windows, with a minimum gap duration. Note how each data key
+has different windows, according to its data distribution.
 
-#### Single global window
+#### 7.2.4. Single global window
```
Comment (tgroh) [inline src/documentation/programming-guide.md]: "The single global window"?

File: src/documentation/programming-guide.md [hunk 7]
```diff
@@ -1372,119 +1923,228 @@ public class MyCustomDataType {
...
+global window default for your `PCollection`.
 
-You can use a single global window if you are working with an unbounded data set (e.g. from a streaming data source) but use caution when applying aggregating transforms such as `GroupByKey` and `Combine`. A single global window with a default trigger generally requires the entire data set to be available before processing, which is not possible with continuously updating data. To perform aggregations on an unbounded `PCollection` that uses global windowing, you should specify a non-default trigger for that `PCollection`.
+You can use a single global window if you are working with an unbounded data set
+(e.g. from a streaming data source) but use caution when applying aggregating
+transforms such as `GroupByKey` and `Combine`. A single global window with a
+default trigger generally requires the entire data set to be available before
+processing, which is not possible with continuously updating data. To perform
+aggregations on an unbounded `PCollection` that uses global windowing, you
+should specify a non-default trigger for that `PCollection`.
```
Comment (tgroh) [inline src/documentation/programming-guide.md]: Maybe we should add (at some later point) a section on a custom window fn?

File: src/documentation/programming-guide.md [hunk 8]
```diff
@@ -1372,119 +1923,228 @@ public class MyCustomDataType {
...
 
-### <a name="setwindowingfunction"></a>Setting your PCollection's windowing function
+### 7.3. Setting your PCollection's windowing function
```
Comment (tgroh) [inline src/documentation/programming-guide.md]: Can we condense the "How to construct a window fn" to the above section, and have this be just how to apply a WindowFn to a PCollection?

(Feel free to JIRA and do later)

File: src/documentation/programming-guide.md [hunk 9]
```diff
@@ -1641,34 +2393,56 @@ The following example shows a billing scenario, and uses both early and late fir
       // Fire on any late data so the bill can be corrected.
       .withLateFirings(AfterPane.elementCountAtLeast(1))
 ```
-
 ```py
   # The Beam SDK for Python does not support triggers.
 ```
 
-##### <a name="default-trigger"></a>**Default Trigger**
+#### 8.1.1. Default trigger
```
Comment (tgroh) [inline src/documentation/programming-guide.md]: **The** default trigger

File: src/documentation/programming-guide.md (hunk 8 above)
Comment (melap) [reply src/documentation/programming-guide.md]: opened JIRA

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/beam_PreCommit_Website_Stage/714/

Jenkins built the site at commit id c5d8da69530aaeae85654c334868abeb7b032b13 with Jekyll and staged it [here](http://apache-beam-website-pull-requests.storage.googleapis.com/320/index.html). Happy reviewing.

Note that any previous site has been deleted. This staged site will be automatically deleted after its TTL expires. Push any commit to the pull request branch or re-trigger the build to get it staged again.

**Comment by tgroh** [comment]
LGTM

**Comment by tgroh** [comment]
@asfgit merge

**Pull request merged**